### ISSUE TITLE
Local Agreement Policy

### ISSUE THREAD

**Issue by sandorkonya** [huggingface/speech-to-speech#72]
Hi there, thank you for the great repo!

I've found recently a method that could probably make the transcription much better and wanted to bring it into your attention:
https://github.com/ufal/whisper_streaming?tab=readme-ov-file#background 

Regards

**Comment by andimarafioti** [comment]
Thank you for sharing! It seems to me like this only becomes an issue for sentences that are 30+ seconds, which is a bit niche for this project. But definitely something to keep in mind for improvements!

**Comment by sandorkonya** [comment]
Hi @andimarafioti, the description there is somewhat missleading, the method was intended for realtime streamed ASR, you might take a look at this 
[explanation video](https://youtu.be/_spinzpEeFM?si=26gKCN14vRgzsvmj&t=202), just a few sec, but it's worth!

In the meanwhile i found the 
[whisperfusion](https://github.com/collabora/WhisperFusion/tree/main?tab=readme-ov-file#whisperfusion
) repo, and their [explanation video](https://www.youtube.com/watch?v=_PnaP0AQJnk) (also a short one) showcases a similar behaviour with constantly changing of the transcribed text mid-sentence. They also use silero's VAD, but in their [code](https://github.com/collabora/WhisperFusion/blob/main/whisper_live/transcriber.py#L384) i see a little more sorcery that takes care of the text segments and punctuation.

Or... maybe i am the one who misses something here... in this HF speech-2-speech repo the VAD takes (by default) 30ms chunks of the audio until no-speech is detected and these chunks will be then concatenated and sent to the STT model which generates the transcript? There is no backchecking / adjusting of generated text like in these 2 examples above, right?

thank you for your time.

**Comment by andimarafioti** [comment]
Ok, you're right, that video is very different to what I thought. 

I'm wondering if there would be a trade-off with performance, but we don't send the sentences from whisper for the LLM until a full sentence is there, so processing the chunks of audio multiple times is not really an issue.

The VAD takes 30 ms chunks of audio until no-speech is detected. Then it sends all of those chunks concatenated + the previous 500 ms before starting to talk, and whisper processes that. I think here we could make everything more efficient if we processed with whisper from the moment that the VAD starts to detect audio. That way, when we stop having speech, we might already have a full-sentence which we could have even sent to the LLM. WDYT @eustlb ? Worth doing?

**Comment by sandorkonya** [comment]
> That way, when we stop having speech, we might already have a full-sentence which we could have even sent to the LLM.

^^ exactly. The longer the sentence is, the more inference time can be saved: since the first part of the sentence is already "confirmed" and only the last group of concatenated chunks are added as "not confirmed", moreover the [condition_on_previous_text](https://utca.knowledgator.com/predictors/openaiwhisperpredictor#init-2) can be used to get better transcriptions for the last chunks.

Since distil-whisper is (much) faster than real time on modern GPU, it should not be a problem when only one STT session is running... probably not even on CPU with multiple threads.

**Comment by sandorkonya** [comment]
Ran some tests on colab with T4. 

Silero's VAD:  `snakers4/silero-vad`
distil-whisper: `primeline/distil-whisper-large-v3-german`
dataset: `flozi00/german-canary-asr-0324`

VAD gives following speech detections for an example sentence, 5 parts:
[{'start': 8224, 'end': 19936},  {'start': 21536, 'end': 31200},  {'start': 32800, 'end': 52704},  {'start': 56864, 'end': 71136},  {'start': 77344, 'end': 84448}]

I sent the whole audio file and the cropped audio slices consecutively:

Complete audio length: 6.3 seconds  |   Execution time: 0.797  seconds | Namensgebend für den Bezirk ist der Lach die.
Part 1  Audio length: 0.732 seconds | Execution time: 0.509  seconds | Nam.
Part 2 Audio length: 0.604 seconds | Execution time: 0.749  seconds  | Hier mein?
Part 3 Audio length: 1.244 seconds | Execution time: 0.446  seconds | Wir haben ihn mit eher? 
Part 4 Audio length: 0.892 seconds | Execution time: 0.495  seconds  | Ist gelach.
Part 5 Audio length: 0.444 seconds | Execution time: 0.505  seconds | Ging.

The transcription of a 6+ sec long audio takes almost as long as the transcription of a <1 sec one.... I did not see that comming.
Note, i sent the cropped audio pieces only, no preceeding audio this is why the transcription results are crappy.

Basicly, sending less than 0.6-0.7 sec long audio yields 1. bad results, 2. the inference time is long, and it should take much less than the length of the audio part itself.
